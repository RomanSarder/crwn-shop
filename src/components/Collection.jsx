import React from 'react'
import styled from 'styled-components'

import CollectionItem from './CollectionItem'

const StyledPreviewCollection = styled.div`
    display: flex;
    flex-direction: column;

    .title {
        font-weight: bold;
        font-size: 2.8rem;
        text-align: center;
        text-transform: capitalize;
    }

    .preview {
        display: flex;
        flex-wrap: wrap;
        justify-content: center;
        gap: 3rem;

        & > * {
            flex-basis: 34rem;
        }
    }
`

export default function Collection({ title, items }) {
    return (
        <StyledPreviewCollection>
            <h2 className="title">{ title }</h2>
            <div className="preview">
                {
                    items
                    .map(function renderCollectionItems({ id, name, imageUrl, price }) {
                        return (
                            <CollectionItem key={id} id={id} name={name} imageUrl={imageUrl} price={price} />
                        )
                    })
                }
            </div>
        </StyledPreviewCollection>
    )
}
