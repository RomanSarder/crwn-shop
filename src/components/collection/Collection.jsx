import React from 'react'
import { useHistory } from 'react-router'
import styled from 'styled-components'

import CollectionItem from '../collection-item/CollectionItem'

const StyledPreviewCollection = styled.div`
    display: flex;
    flex-direction: column;

    .title {
        font-weight: bold;
        font-size: 3.2rem;
        text-align: center;
        text-transform: capitalize;
        cursor: pointer;
        transition: color 0.5s ease-in-out;

        &:hover {
            color: darkgrey;
        }
    }

    .preview {
        display: flex;
        flex-wrap: wrap;
        justify-content: center;
        gap: 3rem;

        & > * {
            flex-basis: 34rem;
        }
    }
`

export function handleCategoryClick (history, title) {
    return () => {
        history.push(`/shop/${title}`)
    }
}

export default function Collection({ title, items, onCategoryClick = handleCategoryClick }) {
    var history = useHistory()

    return (
        <StyledPreviewCollection>
            <h2 className="title" onClick={onCategoryClick(history, title)}>{ title }</h2>
            <div className="preview">
                {
                    items
                    .map(function renderCollectionItems({ id, name, imageUrl, price }) {
                        return (
                            <CollectionItem key={id} id={id} name={name} imageUrl={imageUrl} price={price} />
                        )
                    })
                }
            </div>
        </StyledPreviewCollection>
    )
}
