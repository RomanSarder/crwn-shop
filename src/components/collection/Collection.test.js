import { render } from '@testing-library/react'
import 'jest-styled-components'
import userEvent from '@testing-library/user-event'
import Collection, { handleCategoryClick } from './Collection'

jest.mock('../collection-item/CollectionItem', () => () => (<div>Hello World</div>))

const testTitle = 'Test collection title'
var testCollectionItems = [
    {
        id: 1,
        name: 'test item 1',
        imageUrl: 'test image 1',
        price: 10,
    },
    {
        id: 2,
        name: 'test item 2',
        imageUrl: 'test image 2',
        price: 20
    }
]

it('<Collection /> should properly render with children', () => {
    var { container, getByRole, getAllByText } = render(<Collection items={testCollectionItems} title={testTitle} />)

    expect(container.firstChild).toMatchSnapshot()
    expect(getByRole('heading')).toBeVisible()
    expect(getByRole('heading')).toHaveTextContent(testTitle)
    expect(getAllByText('Hello World').length).toEqual(2)
})

it('<Collection /> should properly trigger handle when title is clicked', () => {
    var handleCategoryClickMock = jest.fn()
    var onCategoryClickMock = () => handleCategoryClickMock
    var { getByRole } = render(<Collection items={testCollectionItems} title={testTitle} onCategoryClick={onCategoryClickMock}/>)

    userEvent.click(getByRole('heading'))

    expect(handleCategoryClickMock).toHaveBeenCalledTimes(1)
})

it('<Collection /> should route to proper category url', () => {
    var historyMock = {
        push: jest.fn()
    }

    handleCategoryClick(historyMock, testTitle)()

    expect(historyMock.push).toHaveBeenCalledWith(`/shop/${testTitle}`)
})