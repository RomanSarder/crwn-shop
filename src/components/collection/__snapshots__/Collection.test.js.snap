// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should properly render with children 1`] = `
<div
  class="sc-bdvvaa dTFZAz"
>
  <h2
    class="title"
  >
    Test collection title
  </h2>
  <div
    class="preview"
  >
    <div>
      Hello World
    </div>
    <div>
      Hello World
    </div>
  </div>
</div>
`;
